aws_ssm_parameter with association name seems to interfere with parameters and vice versa

### Terraform Version
Terraform v0.11.3
+ provider.aws v1.8.0
+ provider.external v1.0.0
+ provider.template v1.0.0

### Affected Resource(s)
- aws_ssm_association

### Terraform Configuration Files
```hcl
resource "aws_ssm_association" "test" {
    association_name = "Test"
    name = "AWS-RunAnsiblePlaybook"
    targets = [
        {
            key = "tag:Test"
            values = ["true"]
        }
    ]
    parameters {
        check = "False"
        playbook = <<EOF
- hosts: all
  become: true
  tasks:
  - name: set debug header
    copy:
      dest: /etc/nginx/conf.d/test.conf
      content: add_header "X-Debug" "Test" always;
      mode: 0644
    notify:
    - reload nginx
  handlers:
  - name: reload nginx
    service:
      name: nginx
      state: reloaded
EOF
    }
}
```

### Expected Behavior
SSM Assocation should have been updated with new ansible script and assocaition name when updated

### Actual Behavior
This varies.

a) On the first run everything goes as expected.

b) If for example the ansible script gets changed, and the terraform is re-run, the parameters will be completely empty

c) If run again (no further changes) then terraform will still detect a change, then do the update, the parameters will be there but the association name will be '-'

d) If run again (no further changes) then terraform will still detect a change, then do the update, the parameters will be empty again, but now the association name will be correct.

e) goto c

I'm having similar problem. I have simple association to run remote shell script.

First run: plan and apply works as expected.
Second plan: shows change in `association_name: "" => "my-association-name"`. Applying this plan and got:`[ERROR] Error updating SSM association: InvalidParameters: Parameter "commands" requires a value.`

The association_name or commands haven't been changed in terraform definition. So far this is blocker for me.

--- a/aws/resource_aws_ssm_association.go
+++ b/aws/resource_aws_ssm_association.go
@@ -198,7 +198,7 @@ func resourceAwsSsmAssocationUpdate(d *schema.ResourceData, meta interface{}) er
 	// AWS creates a new version every time the association is updated, so everything should be passed in the update.
 
 	hasChanges := false
-	
+
 	if d.HasChange("association_name") {
 		hasChanges = true
 	}
@@ -227,22 +227,22 @@ func resourceAwsSsmAssocationUpdate(d *schema.ResourceData, meta interface{}) er
 		if v, ok := d.GetOk("association_name"); ok {
 			associationInput.AssociationName = aws.String(v.(string))
 		}
-	
+
 		if v, ok := d.GetOk("document_version"); ok {
 			associationInput.DocumentVersion = aws.String(v.(string))
 		}
-	
+
 		if v, ok := d.GetOk("schedule_expression"); ok {
 			associationInput.ScheduleExpression = aws.String(v.(string))
 		}
-	
+
 		if v, ok := d.GetOk("parameters"); ok {
 			associationInput.Parameters = expandSSMDocumentParameters(v.(map[string]interface{}))
 		}
-	
+
 		if _, ok := d.GetOk("targets"); ok {
-		associationInput.Targets = expandAwsSsmTargets(d.Get("targets").([]interface{}))
-	}
+			associationInput.Targets = expandAwsSsmTargets(d.Get("targets").([]interface{}))
+		}
 
 		if v, ok := d.GetOk("output_location"); ok {
 			associationInput.OutputLocation = expandSSMAssociationOutputLocation(v.([]interface{}))

--- a/aws/resource_aws_ssm_association.go
+++ b/aws/resource_aws_ssm_association.go
@@ -97,39 +97,39 @@ func resourceAwsSsmAssociationCreate(d *schema.ResourceData, meta interface{}) e
 
 	log.Printf("[DEBUG] SSM association create: %s", d.Id())
 
-	assosciationInput := &ssm.CreateAssociationInput{
+	associationInput := &ssm.CreateAssociationInput{
 		Name: aws.String(d.Get("name").(string)),
 	}
 
 	if v, ok := d.GetOk("association_name"); ok {
-		assosciationInput.AssociationName = aws.String(v.(string))
+		associationInput.AssociationName = aws.String(v.(string))
 	}
 
 	if v, ok := d.GetOk("instance_id"); ok {
-		assosciationInput.InstanceId = aws.String(v.(string))
+		associationInput.InstanceId = aws.String(v.(string))
 	}
 
 	if v, ok := d.GetOk("document_version"); ok {
-		assosciationInput.DocumentVersion = aws.String(v.(string))
+		associationInput.DocumentVersion = aws.String(v.(string))
 	}
 
 	if v, ok := d.GetOk("schedule_expression"); ok {
-		assosciationInput.ScheduleExpression = aws.String(v.(string))
+		associationInput.ScheduleExpression = aws.String(v.(string))
 	}
 
 	if v, ok := d.GetOk("parameters"); ok {
-		assosciationInput.Parameters = expandSSMDocumentParameters(v.(map[string]interface{}))
+		associationInput.Parameters = expandSSMDocumentParameters(v.(map[string]interface{}))
 	}
 
 	if _, ok := d.GetOk("targets"); ok {
-		assosciationInput.Targets = expandAwsSsmTargets(d.Get("targets").([]interface{}))
+		associationInput.Targets = expandAwsSsmTargets(d.Get("targets").([]interface{}))
 	}
 
 	if v, ok := d.GetOk("output_location"); ok {
-		assosciationInput.OutputLocation = expandSSMAssociationOutputLocation(v.([]interface{}))
+		associationInput.OutputLocation = expandSSMAssociationOutputLocation(v.([]interface{}))
 	}
 
-	resp, err := ssmconn.CreateAssociation(assosciationInput)
+	resp, err := ssmconn.CreateAssociation(associationInput)
 	if err != nil {
 		return fmt.Errorf("[ERROR] Error creating SSM association: %s", err)
 	}
@@ -189,7 +189,7 @@ func resourceAwsSsmAssociationRead(d *schema.ResourceData, meta interface{}) err
 func resourceAwsSsmAssocationUpdate(d *schema.ResourceData, meta interface{}) error {
 	ssmconn := meta.(*AWSClient).ssmconn
 
-	log.Printf("[DEBUG] SSM association update: %s", d.Id())
+	log.Printf("[DEBUG] SSM Association update: %s", d.Id())
 
 	associationInput := &ssm.UpdateAssociationInput{
 		AssociationId: aws.String(d.Get("association_id").(string)),
@@ -230,7 +230,7 @@ func resourceAwsSsmAssocationUpdate(d *schema.ResourceData, meta interface{}) er
 func resourceAwsSsmAssociationDelete(d *schema.ResourceData, meta interface{}) error {
 	ssmconn := meta.(*AWSClient).ssmconn
 
-	log.Printf("[DEBUG] Deleting SSM Assosciation: %s", d.Id())
+	log.Printf("[DEBUG] Deleting SSM Association: %s", d.Id())
 
 	params := &ssm.DeleteAssociationInput{
 		AssociationId: aws.String(d.Get("association_id").(string)),

--- a/aws/resource_aws_ssm_association.go
+++ b/aws/resource_aws_ssm_association.go
@@ -13,7 +13,7 @@ func resourceAwsSsmAssociation() *schema.Resource {
 	return &schema.Resource{
 		Create: resourceAwsSsmAssociationCreate,
 		Read:   resourceAwsSsmAssociationRead,
-		Update: resourceAwsSsmAssocationUpdate,
+		Update: resourceAwsSsmAssociationUpdate,
 		Delete: resourceAwsSsmAssociationDelete,
 
 		MigrateState:  resourceAwsSsmAssociationMigrateState,
@@ -186,7 +186,7 @@ func resourceAwsSsmAssociationRead(d *schema.ResourceData, meta interface{}) err
 	return nil
 }
 
-func resourceAwsSsmAssocationUpdate(d *schema.ResourceData, meta interface{}) error {
+func resourceAwsSsmAssociationUpdate(d *schema.ResourceData, meta interface{}) error {
 	ssmconn := meta.(*AWSClient).ssmconn
 
 	log.Printf("[DEBUG] SSM Association update: %s", d.Id())

--- a/aws/resource_aws_ssm_association.go
+++ b/aws/resource_aws_ssm_association.go
@@ -195,28 +195,66 @@ func resourceAwsSsmAssocationUpdate(d *schema.ResourceData, meta interface{}) er
 		AssociationId: aws.String(d.Get("association_id").(string)),
 	}
 
+	// AWS creates a new version every time the association is updated, so everything should be passed in the update.
+
+	hasChanges := false
+	
 	if d.HasChange("association_name") {
-		associationInput.AssociationName = aws.String(d.Get("association_name").(string))
+		hasChanges = true
 	}
 
-	if d.HasChange("schedule_expression") {
-		associationInput.ScheduleExpression = aws.String(d.Get("schedule_expression").(string))
+	if d.HasChange("instance_id") {
+		hasChange = true
 	}
 
 	if d.HasChange("document_version") {
-		associationInput.DocumentVersion = aws.String(d.Get("document_version").(string))
+		hasChanges = true
+	}
+
+	if d.HasChange("schedule_expression") {
+		hasChanges = true
 	}
 
 	if d.HasChange("parameters") {
-		associationInput.Parameters = expandSSMDocumentParameters(d.Get("parameters").(map[string]interface{}))
+		hasChanges = true
 	}
 
 	if d.HasChange("output_location") {
-		associationInput.OutputLocation = expandSSMAssociationOutputLocation(d.Get("output_location").([]interface{}))
+		hasChanges = true
 	}
 
 	if d.HasChange("targets") {
+		hasChanges = true
+	}
+
+	if hasChanges {
+		if v, ok := d.GetOk("association_name"); ok {
+			associationInput.AssociationName = aws.String(v.(string))
+		}
+
+		if v, ok := d.GetOk("instance_id"); ok {
+			associationInput.InstanceId = aws.String(v.(string))
+		}
+	
+		if v, ok := d.GetOk("document_version"); ok {
+			associationInput.DocumentVersion = aws.String(v.(string))
+		}
+	
+		if v, ok := d.GetOk("schedule_expression"); ok {
+			associationInput.ScheduleExpression = aws.String(v.(string))
+		}
+	
+		if v, ok := d.GetOk("parameters"); ok {
+			associationInput.Parameters = expandSSMDocumentParameters(v.(map[string]interface{}))
+		}
+	
+		if _, ok := d.GetOk("targets"); ok {
 		associationInput.Targets = expandAwsSsmTargets(d.Get("targets").([]interface{}))
+	}
+
+		if v, ok := d.GetOk("output_location"); ok {
+			associationInput.OutputLocation = expandSSMAssociationOutputLocation(v.([]interface{}))
+		}
 	}
 
 	_, err := ssmconn.UpdateAssociation(associationInput)

--- a/aws/resource_aws_ssm_association.go
+++ b/aws/resource_aws_ssm_association.go
@@ -203,10 +203,6 @@ func resourceAwsSsmAssocationUpdate(d *schema.ResourceData, meta interface{}) er
 		hasChanges = true
 	}
 
-	if d.HasChange("instance_id") {
-		hasChange = true
-	}
-
 	if d.HasChange("document_version") {
 		hasChanges = true
 	}
@@ -230,10 +226,6 @@ func resourceAwsSsmAssocationUpdate(d *schema.ResourceData, meta interface{}) er
 	if hasChanges {
 		if v, ok := d.GetOk("association_name"); ok {
 			associationInput.AssociationName = aws.String(v.(string))
-		}
-
-		if v, ok := d.GetOk("instance_id"); ok {
-			associationInput.InstanceId = aws.String(v.(string))
 		}
 	
 		if v, ok := d.GetOk("document_version"); ok {

